### PR TITLE
Support themed icons for 3rd party icon packs

### LINKED ISSUE #2678: [FEATURE REQUEST] Ability to assign Lawnicon as custom icon
When customizing an app, the ability to choose an adaptive Lawnicon would be very useful.

### LINKED ISSUE #3007: Ability to change themed icons from Lawnicons
### Describe the feature

To be able to change what icon appears when themed icons is enabled using Lawnicons.

### Why would it be useful to add?

Certain apps aren't themed but have similar icons to apps that do (i.e YT Music Revanced doesn't have a themed icon, but could change that to the official YT Music app)

### Did the feature exist in Lawnchair v2? (Play Store version)

No

### Example(s)

Screenshot_20221006_205142
YT Music not themed
Screenshot_20221006_205201
Add option within here

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by naveenccmsd**
## Description
Here I'm planning to cover three feature 

- In app Customisiation for themed Icons(lawnicons)
- LawnIcons as Iconpack (So Lawnicon can be used for icon pack , We can still use lawnicons as Monochromic icons in lwanchair, may be same can be used for othe rlaucnhers too)
- Add Themed Icons support for other icons pack as well apart from lawncions.

Sample:

Lawnicons

https://user-images.githubusercontent.com/7315975/200399398-99cb333b-9f0c-4b06-8a47-bc7484ba0726.mp4

App customisiation from 3rd party icons

https://user-images.githubusercontent.com/7315975/200399425-e024bdc1-d6be-4f7a-aada-97589c62245e.mp4

I have tested with own version of lawnicons (I copied lawicons file to other open source icon pack) to test .

https://github.com/naveenccmsd/lawnicons/releases/download/test/app-debug.apk

Based on this PR  I can work Lawnicons as iconpack

Closes #3007
Closes #2678

 :white_check_mark:  New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: lawnchair/res/values/config.xml
Comment (naveenccmsd) [inline lawnchair/res/values/config.xml:35]: Intrestingly lawnchair works well with icon packs that were made from svg with trasparanet background and BW foreground. Found below icon packs works well. So added here let me know if this can be added

File: lawnchair/src/app/lawnchair/icons/IconPackProvider.kt
Comment (Goooler) [inline lawnchair/src/app/lawnchair/icons/IconPackProvider.kt]:
```suggestion
    private val themedIconPacks = context.resources.getStringArray(R.array.themed_icon_packs)
```

File: lawnchair/src/app/lawnchair/icons/IconPackProvider.kt
Comment (Goooler) [inline lawnchair/src/app/lawnchair/icons/IconPackProvider.kt]:
```suggestion
        if (
            context.isThemedIconsEnabled() && iconEntry.packPackageName in themedIconPacks
        ) {
            val themedColors: IntArray = ThemedIconDrawable.getThemedColors(context)
            val res = context.packageManager.getResourcesForApplication(iconEntry.packPackageName)
            @SuppressLint("DiscouragedApi")
            val resId = res.getIdentifier(iconEntry.name, "drawable", iconEntry.packPackageName)
            val bg: Drawable = ColorDrawable(themedColors[0])
            val td = ThemedIconDrawable.ThemeData(res, iconEntry.packPackageName, resId)
            val fg = td.wrapDrawable(drawable, 0)
            return if (fg is AdaptiveIconDrawable) {
                val foregroundDr: Drawable = fg.getForeground()
                foregroundDr.setTint(themedColors[1])
                CustomAdaptiveIconDrawable(bg, foregroundDr)
            } else {
                val iconFromPack = InsetDrawable(drawable, .3f)
                iconFromPack.setTint(themedColors[1])
                td.wrapDrawable(CustomAdaptiveIconDrawable(bg, iconFromPack), 0)
            }
        }
```

**Comment by Goooler** [comment]
Need to import `SuppressLint`.

File: lawnchair/res/values/config.xml
Comment (Goooler) [inline lawnchair/res/values/config.xml]: Please revert this unrelated line.

**Comment by Goooler** [comment]
I've merged https://github.com/LawnchairLauncher/platform_frameworks_libs_systemui/pull/11 now, please resync the submodule to https://github.com/LawnchairLauncher/platform_frameworks_libs_systemui/commit/34742ba4c94c9d962668c6732428aaa8c84581cd.

**Comment by Goooler** [comment]
You should resync to 34742ba4c94c9d962668c6732428aaa8c84581cd instead of e34a52c06299ab1288d4f6086d146a3e39ab91a0

File: lawnchair/res/values/config.xml
Comment (yasanglass) [reply lawnchair/res/values/config.xml:35]: While this is great, having to keep a static list of packages up to date can be very hard, especially since there are tons of icon packs. If we manage to allow users to modify the list on runtime it would be much better, especially for the maintainers.

Comment (Goooler) [reply lawnchair/res/values/config.xml:35]: Good idea, so we need a selector to select icon packs from apps list?

Comment (naveenccmsd) [reply lawnchair/res/values/config.xml:35]: But one problem is icons might get break if themed enabled for not supported icon pack. 

If this can be decided by user then we don't need this check whether packages installed to enable _themed icons_ option .

So user will decide if they can enable  _themed icons_  options for the icon pack they selected

Comment (naveenccmsd) [reply lawnchair/res/values/config.xml:35]: If themed enabled for unsupported icon pack
WhatsApp Image 2022-11-08 at 4 31 23 PM

in this can user need to turn off themed option

Comment (naveenccmsd) [reply lawnchair/res/values/config.xml:35]: https://user-images.githubusercontent.com/7315975/200548360-716d3cb1-2c38-44cb-846e-207341f16441.mp4

Comment (naveenccmsd) [reply lawnchair/res/values/config.xml:35]: How about having static list with lawnchair and whatever tested will be added here with PR. Having an advanced option, may be in experimetal feature that enables it for all icon packs too.

Comment (yasanglass) [reply lawnchair/res/values/config.xml:35]: We can warn the user that not all icon packs support this feature. We just have to make sure it's clear enough so we don't get tons of messages about issues.

About having a static list, I strongly disagree with it at least if it's locally stored. Because not only do we have to update the list on the code, we have to push updates frequently just to let users use the icon packs they want. Basically a bad experience for both sides.

I think the final feature would be ideal if:
- The local static list can be updated from a remote source.
- The user can select any icon pack to be used with proper warnings etc but it probably should still be an experimental feature for more advanced users.

Although... this does sound like lots of more work, especially on the UI UX aspects. Maybe it would be better if we just merge the current PR with the core feature and then add additional QoL improvements with time.

**Comment by yasanglass** [comment]
Is the PR final? Will the [additional QoL changes](https://github.com/LawnchairLauncher/lawnchair/pull/3091#discussion_r1016498818) be in future PRs? Want to know if I should review and merge it yet or not.

**Comment by naveenccmsd** [comment]
> t to know if I should review and merge it yet or not.

@yasandev , yes we can have the https://github.com/LawnchairLauncher/lawnchair/pull/3091#discussion_r1016498818 in future PR.

**Review by yasanglass — APPROVED** [review]

**Review by Goooler — APPROVED** [review]

**Comment by yasanglass** [comment]
@naveenccmsd thank you for contributing to Lawnchair!

**Pull request merged**
